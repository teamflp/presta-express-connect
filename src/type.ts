// types.ts
import { RootState } from './store/store'; // Assurez-vous que le chemin est correct

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;