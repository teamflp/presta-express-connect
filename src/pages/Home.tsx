
import Header from '../components/Header/Header';
import Footer from '../components/Footer/Footer';
import Cities from '../components/City/Cities';
import FaqAccordion from '../components/FAQ/Faqs';
import Infos from '../components/Newsletter/Infos';
import Newsletter from '../components/Newsletter/Newsletter';
import CardArguments from '../components/Arguments/Arguments';
import LoremsPresta from '../components/Lorem/Lorems';
import ArtisanByLetter from '../components/SectionProf/ArtisanByLetter';
import CategoriesArtisans from '../components/Artisant/CategoriesArtisans';
import RegistrationArtisan from '../components/SectionProf/RegistrationArtisan';
import City from '../components/City/City';


function Home() {
  return (
    <div className="App">
      <Header />
      <div className="my-8">
        <CardArguments />
      </div>
      <div className="container my-15 container-carousel">
        <CategoriesArtisans />
      </div>
      <div className="container px-0 my-15">
        <div className="row my-5 d-flex justify-content-between">
          <div className="col-12 col-md-5">
            <ArtisanByLetter />
          </div>
          <div className="col-12 col-md-7">
            <RegistrationArtisan />
          </div>
        </div>
      </div>
      <div className="my-15">
        <LoremsPresta />
      </div>
      <div className="my-15">
        <City />
      </div>
      <div className="container px-0 my-15">
        <div className="row my-5 d-flex justify-content-between">
          <div className="col-12 col-md-8">
            <Newsletter />
          </div>
          <div className="col-12 col-md-4">
            <Infos />
          </div>
        </div>
      </div>
      <div className="container my-3 my-md-5">
        <FaqAccordion />
      </div>
      <Footer />
    </div>
  );
}

export default Home;

