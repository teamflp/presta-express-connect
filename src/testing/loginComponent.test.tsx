import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import LoginComponent from '../components/User/loginComponent';
import usersReducer, { userLogin } from '../store/features/users/usersSlice';
import authReducer from '../store/features/Authentification/AuthSlice';
import { Route, Routes } from 'react-router-dom';
import { MemoryRouter } from 'react-router-dom';
import { AppDispatch } from '../type';
import Home from '../pages/Home';

// Nettoyage du localStorage avec et après chaque test
beforeEach(() => {
  localStorage.clear();
});

afterEach(() => {
  localStorage.clear();
});

/// Mock de useNavigate
const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

// Configurez le store avec uniquement les reducers nécessaires
const store = configureStore({
  reducer: {
    users: usersReducer, // Reducer utilisé par LoginComponent
    auth: authReducer, // Reducer utilisé pour gérer l'authentification
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: ['persist/PERSIST'],
      },
    }),
});

test('rendu correct', () => {
  const { container } = render(
    <Provider store={store}>
      <MemoryRouter>
        <LoginComponent />
      </MemoryRouter>
    </Provider>
  );
  expect(container).toMatchSnapshot();
});

test('rendu avec champs email et password vide', () => {
  const { getByPlaceholderText } = render(
    <Provider store={store}>
      <MemoryRouter>
        <LoginComponent />
      </MemoryRouter>
    </Provider>
  );

  const emailInput = getByPlaceholderText('Email') as HTMLInputElement;
  const passwordInput = getByPlaceholderText(
    'Mot de passe'
  ) as HTMLInputElement;

  expect(emailInput.value).toBe('');
  expect(passwordInput.value).toBe('');
});

test("rendu avec le bouton 'Se connecter' disabled si les champs email et password sont vide", () => {
  const { getByPlaceholderText, getByRole } = render(
    <Provider store={store}>
      <MemoryRouter>
        <LoginComponent />
      </MemoryRouter>
    </Provider>
  );

  const loginButton = getByRole('button', { name: 'Se connecter' });

  const emailInput = getByPlaceholderText('Email') as HTMLInputElement;
  const passwordInput = getByPlaceholderText(
    'Mot de passe'
  ) as HTMLInputElement;

  expect(emailInput.value).toBe('');
  expect(passwordInput.value).toBe('');
  expect(loginButton).toBeDisabled();
});

test("mise à jour de l'état 'se souvenir de moi' lorsque la checkbox est cochée", () => {
  const { getByLabelText } = render(
    <Provider store={store}>
      <MemoryRouter>
        <LoginComponent />
      </MemoryRouter>
    </Provider>
  );

  const rememberMeCheckbox = getByLabelText('Se souvenir de moi');

  // Vérifie si la checkbox est cochée au début
  expect(rememberMeCheckbox).not.toBeChecked();

  // Simule un clic sur la checkbox
  fireEvent.click(rememberMeCheckbox);

  // Vérifie que la checkbox est maintenant cochée
  expect(rememberMeCheckbox).toBeChecked();

  // Simule un autre click sur la checkbox
  fireEvent.click(rememberMeCheckbox);

  // Vérifie maintenant si la checkbox n'est plus cochée
  expect(rememberMeCheckbox).not.toBeChecked();
});

test('gère correctement un échec de connexion', async () => {
  // Simulez l'échec de l'authentification
  const user = { email: 'user1@test.com', password: 'wrongpassword' };

  // Définit les valeurs dans le localStorage
  localStorage.setItem('userEmail', JSON.stringify(user.email));
  localStorage.setItem('userPassword', JSON.stringify(user.password));

  const { getByPlaceholderText, getByText } = render(
    <Provider store={store}>
      <MemoryRouter>
        <LoginComponent />
      </MemoryRouter>
    </Provider>
  );

  const emailInput = getByPlaceholderText('Email');
  const passwordInput = getByPlaceholderText('Mot de passe');

  // Simule la saisie des informations de connexion
  fireEvent.change(emailInput, { target: { value: user.email } });
  fireEvent.change(passwordInput, { target: { value: user.password } });

  // Simule la soumission du formulaire
  fireEvent.click(getByText('Se connecter'));

  // Attendre que les messages d'erreur apparaissent
  await waitFor(() => {
    expect(getByText('Identifiants incorrects')).toBeInTheDocument();
  });
});

test('supprime les informations de localstorage quand la checkbox est décochée', async () => {
  const user = { email: 'user1@test.com', password: '12345' };

  // Définit les valeurs dans le localStorage
  localStorage.setItem('userEmail', JSON.stringify(user.email));
  localStorage.setItem('userPassword', JSON.stringify(user.password));

  const { getByLabelText, getByText } = render(
    <Provider store={store}>
      <MemoryRouter>
        <LoginComponent />
      </MemoryRouter>
    </Provider>
  );

  const rememberMeCheckbox = getByLabelText(
    'Se souvenir de moi'
  ) as HTMLInputElement;

  // Vérifiez l'état initial de la checkbox
  expect(rememberMeCheckbox.checked).toBe(true);

  // Décoche la checkbox
  fireEvent.click(rememberMeCheckbox);

  // Simule la soumission du formulaire pour déclencher la suppression
  fireEvent.click(getByText('Se connecter'));

  // Attendre que les valeurs du localStorage soient mises à jour
  await waitFor(() => {
    expect(localStorage.getItem('userEmail')).toBeNull();
    expect(localStorage.getItem('userPassword')).toBeNull();
  });
});

test('gère la connexion réussie', async () => {
  const user = { email: 'user1@test.com', password: '12345' };

  // Mocker le dispatch pour qu'il retourne une action réussie
  const dispatch = jest.spyOn(
    store,
    'dispatch'
  ) as jest.MockedFunction<AppDispatch>;
  dispatch.mockResolvedValue({
    type: userLogin.fulfilled.type,
    payload: { email: user.email },
  } as unknown); // Remplacer any par unknown pour contourner les erreurs de type

  // Initialiser le mockNavigate
  mockNavigate.mockImplementation(() => {});

  const { getByPlaceholderText, getByText, getByLabelText } = render(
    <Provider store={store}>
      <MemoryRouter>
        <Routes>
          <Route path="/" element={<LoginComponent />} />
          <Route path="/home" element={<Home />} />
        </Routes>
      </MemoryRouter>
    </Provider>
  );

  const emailInput = getByPlaceholderText('Email');
  const passwordInput = getByPlaceholderText('Mot de passe');
  const rememberMeCheckbox = getByLabelText(
    'Se souvenir de moi'
  ) as HTMLInputElement;

  // Simule la saisie des informations de connexion
  fireEvent.change(emailInput, { target: { value: user.email } });
  fireEvent.change(passwordInput, { target: { value: user.password } });

  // Coche la checkbox
  fireEvent.click(rememberMeCheckbox);

  // Simule la soumission du formulaire
  fireEvent.click(getByText('Se connecter'));

  // Attendre que la redirection se produise
  await waitFor(() => {
    expect(mockNavigate).toHaveBeenCalledWith('/home');
  });

  // Vérifiez que les informations sont stockées dans le localStorage
  expect(localStorage.getItem('userEmail')).toBe(JSON.stringify(user.email));
  expect(localStorage.getItem('userPassword')).toBe(
    JSON.stringify(user.password)
  );
});
