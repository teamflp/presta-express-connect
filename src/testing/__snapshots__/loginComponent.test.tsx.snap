// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendu correct 1`] = `
<div>
  <div
    class="containerLogin"
  >
    <div
      class="loginDiv"
    >
      <div
        class="divIconLogin"
      >
        <img
          id="iconArtisanLogin"
          src="test-file-stub"
        />
      </div>
      <h2>
        Connectez-vous et profitez de nos fonctionnalités
      </h2>
      <div
        class="inputLoginDiv"
      >
        <input
          placeholder="Email"
          type="email"
          value=""
        />
        <br />
        <input
          placeholder="Mot de passe"
          type="password"
          value=""
        />
      </div>
      <div
        class="checkboxLoginDiv"
      >
        <input
          id="rememberMe"
          type="checkbox"
        />
        <label
          for="rememberMe"
        >
          Se souvenir de moi
        </label>
      </div>
      <p>
        <a
          href="/"
          id="mdpOublie"
        >
          Mot de passe oublié ?
        </a>
      </p>
      <button
        class="buttonLogin"
        disabled=""
      >
        Se connecter
        <span>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M438.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L338.8 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l306.7 0L233.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
